core/src/main/java/org/elasticsearch/search/aggregations/support/SingleValuesSourceAggregatorBuilder.java

```diff
@@ -0,0 +1,216 @@
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.InternalAggregation.Type;
+import org.joda.time.DateTimeZone;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ *
+ */
+public abstract class SingleValuesSourceAggregatorBuilder<VS extends ValuesSource, AB extends SingleValuesSourceAggregatorBuilder<VS, AB>>
+    extends ValuesSourceAggregatorBuilder<VS, AB> {
+
+    public static abstract class LeafOnly<VS extends ValuesSource, AB extends SingleValuesSourceAggregatorBuilder<VS, AB>>
+        extends SingleValuesSourceAggregatorBuilder<VS, AB> {
+
+        protected LeafOnly(String name, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) {
+            super(name, type, valuesSourceType, targetValueType);
+        }
+
+        /**
+         * Read an aggregation from a stream that does not serialize its targetValueType. This should be used by most subclasses.
+         */
+        protected LeafOnly(StreamInput in, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) throws IOException {
+            super(in, type, valuesSourceType, targetValueType);
+        }
+
+        /**
+         * Read an aggregation from a stream that serializes its targetValueType. This should only be used by subclasses that override
+         * {@link #serializeTargetValueType()} to return true.
+         */
+        protected LeafOnly(StreamInput in, Type type, ValuesSourceType valuesSourceType) throws IOException {
+            super(in, type, valuesSourceType);
+        }
+
+        @Override
+        public AB subAggregations(Builder subFactories) {
+            throw new AggregationInitializationException("Aggregator [" + name + "] of type [" + type + "] cannot accept sub-aggregations");
+        }
+    }
+
+    private String field;
+    private Script script;
+
+    protected SingleValuesSourceAggregatorBuilder(String name, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) {
+        super(name, type, valuesSourceType, targetValueType);
+    }
+
+    /**
+     * Read an aggregation from a stream that does not serialize its targetValueType. This should be used by most subclasses.
+     */
+    protected SingleValuesSourceAggregatorBuilder(StreamInput in, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType)
+        throws IOException {
+        super(in, type, valuesSourceType, targetValueType);
+    }
+
+    /**
+     * Read an aggregation from a stream that serializes its targetValueType. This should only be used by subclasses that override
+     * {@link #serializeTargetValueType()} to return true.
+     */
+    protected SingleValuesSourceAggregatorBuilder(StreamInput in, Type type, ValuesSourceType valuesSourceType) throws IOException {
+        super(in, type, valuesSourceType);
+    }
+
+    /**
+     * Read from a stream.
+     */
+    protected void read(StreamInput in) throws IOException {
```

Comment:
I think the accepted convention now is to read the stream in a Constructor that takes StreamInput as a constructor arg